server/src/com/thoughtworks/go/server/service/GoConfigService.java

```diff
@@ -581,17 +585,33 @@ public void updateAgentAttributes(String uuid, String userName, String hostname,
             command.addCommand(GoConfigDao.updateApprovalStatus(uuid, false));
         }
 
-        if (enable.isFalse()){
+        if (enable.isFalse()) {
             command.addCommand(GoConfigDao.updateApprovalStatus(uuid, true));
         }
 
         if (hostname != null) {
             command.addCommand(new GoConfigDao.UpdateAgentHostname(uuid, hostname, userName));
         }
+
         if (resources != null) {
             command.addCommand(new GoConfigDao.UpdateResourcesCommand(uuid, new Resources(resources)));
         }
 
+        if (environments != null) {
+            Set<String> existingEnvironments = cruiseConfig().getEnvironments().environmentsForAgent(uuid);
+            Set<String> newEnvironments = new HashSet<>(Arrays.asList(environments.split(",")));
+
+            Set<String> environmentsToRemove = Sets.difference(existingEnvironments, newEnvironments);
+            Set<String> environmentsToAdd = Sets.difference(newEnvironments, existingEnvironments);
+
+            for (String environmentToRemove : environmentsToRemove) {
+                command.addCommand(new GoConfigDao.ModifyEnvironmentCommand(uuid, environmentToRemove, TriStateSelection.Action.remove));
+            }
+
+            for (String environmentToAdd : environmentsToAdd) {
+                command.addCommand(new GoConfigDao.ModifyEnvironmentCommand(uuid, environmentToAdd, TriStateSelection.Action.add));
```

Comment:
If the environment does not exist `GoConfigDao::ModifyEnvironmentCommand` will throw a `NoSuchEnvironmentException` exception via `EnvironmentsConfig#named`. This is already tested well.

This exception will be caught in `AgentService#updateAgentAttributes` along with the relevant message. As of now the server responds with a `500`, but that is something I'll change to respond with a `422` instead. Here's the response as of now -

```shell
$ curl -X PATCH \
    -H "Accept: application/vnd.go.cd.v2+json" \
    -H "Content-Type: application/json" \
    -d '{
      "environments": ["foo"]
    }' \
    'http://localhost:8153/go/api/agents/10c357bf-f7ae-42ff-a61f-fe3a8575a9a7'

{
  "message": "Updating agents failed:  { Environment [foo] does not exist. }"
}
```